File: aws-examples/elixir/minimal-s3-demo/test/support/head_bucket_http_mock.ex
defmodule HeadBucketHttpMock do
  @moduledoc false
  @behaviour RuntimeHttpClient

  alias RuntimeHttpClient.Request
  alias RuntimeTypes.HttpResponse

  @responses_key {__MODULE__, :responses}
  @count_key {__MODULE__, :count}

  def reset(responses) when is_list(responses) do
    :persistent_term.put(@responses_key, responses)
    :persistent_term.put(@count_key, 0)
    :ok
  end

  def call_count do
    :persistent_term.get(@count_key, 0)
  end

  @impl RuntimeHttpClient
  def request(%Request{method: :head, url: url}) do
    if String.ends_with?(url, "/my-bucket") do
      count = :persistent_term.get(@count_key, 0)
      :persistent_term.put(@count_key, count + 1)

      case Enum.at(:persistent_term.get(@responses_key, []), count) do
        {:status, status} ->
          {:ok, %HttpResponse{status: status, headers: [], body: ""}}

        nil ->
          {:error, {:no_mock_response, count}}
      end
    else
      {:error, {:unexpected_request, url}}
    end
  end

  def request(%Request{} = req) do
    {:error, {:unexpected_request, req}}
  end
end

File: aws-examples/erlang/minimal-s3-demo/test/head_bucket_http_mock.erl
-module(head_bucket_http_mock).
-behaviour(runtime_http_client).

-include("runtime_types.hrl").
-include("runtime_http_client.hrl").

-export([request/1, reset/1, call_count/0]).

-define(RESPONSES_KEY, {head_bucket_http_mock, responses}).
-define(COUNT_KEY, {head_bucket_http_mock, count}).

reset(Responses) when is_list(Responses) ->
    persistent_term:put(?RESPONSES_KEY, Responses),
    persistent_term:put(?COUNT_KEY, 0),
    ok.

call_count() ->
    persistent_term:get(?COUNT_KEY, 0).

request(#http_client_request{method = head, url = Url}) ->
    case binary:match(Url, <<"/my-bucket">>) of
        nomatch ->
            {error, {unexpected_request, Url}};
        _ ->
            N = call_count(),
            persistent_term:put(?COUNT_KEY, N + 1),
            Responses = persistent_term:get(?RESPONSES_KEY, []),
            case lists:nth(N + 1, Responses ++ [no_response]) of
                {status, Status} ->
                    {ok, #http_response{status = Status, headers = [], body = <<>>}};
                no_response ->
                    {error, {no_mock_response, N}}
            end
    end;
request(#http_client_request{} = Req) ->
    {error, {unexpected_request, Req}}.
